src/core/client_config/resolvers/zookeeper_resolver.c

```diff
@@ -0,0 +1,452 @@
+        port = (char *)cur->value;
+        if (host != NULL)
+          break;
+      }
+    }
+    if (host != NULL && port != NULL) {
+      address = gpr_malloc(GRPC_MAX_SOCKADDR_SIZE);
+      memset(address, 0, GRPC_MAX_SOCKADDR_SIZE);
+      strcat(address, host);
+      strcat(address, ":");
+      strcat(address, port);
+    }
+    grpc_json_destroy(json);
+  } 
+
+  return address;
+}
+
+static void zookeeper_get_children_node_completion(int rc, const char *value, int value_len, 
+                                          const struct Stat *stat, const void *arg) {
+  char *address = NULL;
+  zookeeper_resolver *r = (zookeeper_resolver *)arg;
+
+  if (rc) {
+    gpr_log(GPR_ERROR, "Error in getting a child node of %s", r->name);
+    return;
+  }
+
+  address = zookeeper_parse_address((char *)value, value_len);
+  if (address != NULL) {
+    /* Further resolve address by DNS */
+    grpc_resolve_address(address, NULL, zookeeper_dns_resolved, r);
+    gpr_free(address);
+  } else {
+    gpr_mu_lock(&r->mu);
+    r->resolved_total--;
+    if (r->resolved_num == r->resolved_total) {
+      gpr_mu_unlock(&r->mu);
+      zookeeper_on_resolved(r, r->resolved_addrs);
+    } else {
+      gpr_mu_unlock(&r->mu);
+    }
+  }
+}
+
+static void zookeeper_get_children_completion(int rc, const struct String_vector *children, 
+                                        const void *arg) {
+  int status;
+  char path[GRPC_MAX_ZOOKEEPER_BUFFER_SIZE];
+  int i;
+  zookeeper_resolver *r = (zookeeper_resolver *)arg;
+
+  if (rc) {
+    gpr_log(GPR_ERROR, "Error in getting zookeeper children of %s", r->name);
+    return;
+  }
+
+  if (children->count == 0) {
+    gpr_log(GPR_ERROR, "Fail to resolve zookeeper address %s", r->name);
+    return;
+  }
+
+  r->resolved_addrs = gpr_malloc(sizeof(grpc_resolved_addresses));
+  r->resolved_addrs->addrs = NULL;
+  r->resolved_addrs->naddrs = 0;
+  r->resolved_total = children->count;
+
+  for (i = 0; i < children->count; i++) {
+    memset(path, 0, GRPC_MAX_ZOOKEEPER_BUFFER_SIZE);
+    strcat(path, r->name);
+    strcat(path, "/");
+    strcat(path, children->data[i]);
+    status = zoo_aget(r->zookeeper_handle, path, GRPC_ZOOKEEPER_WATCH, 
+                    zookeeper_get_children_node_completion, r);
+    if (status)
+      gpr_log(GPR_ERROR, "Error in getting zookeeper node %s", path);
+  }
+}
+
+static void zookeeper_get_node_completion(int rc, const char *value, int value_len, 
+                                          const struct Stat *stat, const void *arg) {
+  int status;
+  char *address = NULL;
+  zookeeper_resolver *r = (zookeeper_resolver *)arg;
+
+  r->resolved_addrs = NULL;
+  r->resolved_total = 0;
+  r->resolved_num = 0;
+
+  if (rc) {
+    gpr_log(GPR_ERROR, "Error in getting zookeeper node %s", r->name);
+    return;
+  }
+
+  /* If zookeeper node of path r->name does not have address (i.e. service node),
+     get its children */
+  address = zookeeper_parse_address((char *)value, value_len);
```

Comment:
same as before; don't cast away the const: change the function signature to take a char* value